### PR TITLE
Domain picker: Render unavailable domains when the user searches for a FQDN

### LINKED ISSUE #41066: Domain picker: searching for FQDNs does not always return the TLD specified
In the domain picker, when you're searching for a certain domain name including the TLD such as 
`blogadyblogblagadoo.blog`, the search results should include the domain name that we searched for. In the attached screenshot, you aren't given the option to select the .blog TLD.

Screen Shot 2020-04-14 at 7 49 04 AM

#### Steps to reproduce
On https://wordpress.com/new fill out the details until the domain picker is enabled and open it.
Search for a domain including the tld part e.g. `blogadyblogblagadoo.blog`

### LINKED ISSUE #41219: Domain Picker: when explicitly searching for a domain, check if it's available
**Update (22.12.2020)**

* We can have the availability check directly inside DomainPicker. Tracked in https://github.com/Automattic/wp-calypso/issues/48634
* Add also an exact `{SUBDOMAIN_NAME}.*.blog` free subdomain option to the suggestions list if the search query includes `'.blog'`.

___

pbAok1-1MG-p2#comment-3587

Screenshot 2020-12-22 at 17 56 47

___

via https://github.com/Automattic/wp-calypso/issues/40971#issuecomment-614828121

> [...] when the user explicitly searches for a FQDN (_Fully Qualified Domain Name_), including a subdomain. That means they have a specific domain in mind and we should show its availability. If it's available, we should be sure to communicate that to the user - currently we simply insert it as the first search result:
> 
> Screenshot 2020-04-16 at 18 35 16
> 
> (since the user _might_ be interested in other suggestions, which might be more attractive)
> 
> But if the domain is unavailable, we should guide them on what are their options. For example, this domain is registered, but the user can transfer it in:
> 
> Screenshot 2020-04-16 at 18 36 52

Different possible domain states are listed at https://github.com/Automattic/wp-calypso/blob/f2c9d99b07e14ea938829e9b2e5b31b6fb388a90/client/lib/domains/registration/availability-messages.js

Related: https://github.com/Automattic/wp-calypso/issues/41220

Design: 

image

See details in pbAok1-Hb-p2#comment-1472

### Acceptance criteria
- [x] When the users search for a FQDN that is unavailable, list their query as an unavailable suggestion

### Follow up
- [ ] Explain why it is unavailable. The list of explanations is https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/domains/registration/availability-messages.js#L40-L390.

### PR DESCRIPTION
**Pull request by alshakero**
#### Changes proposed in this Pull Request

* This PR is the first step to fix [this](https://github.com/Automattic/wp-calypso/issues/41219) issue. 
* This preliminary first step only adds the FQDN to the results with unavailable status.
* The next step should be explaining why the domain isn't available by querying the availability endpoint. This one is a bit of a rabbit hole as we'll have to package-ize [these.](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/domains/registration/availability-messages.js#L40-L390) 

#### Testing instructions

1. go to [/new](https://calypso.live/new?branch=update/show-fqdn-queries).
2. reach the domain step.
3. search for an unavailable FQDN (eg google.com), the first item should be your query but marked unavailable.
4. search for an available FQDN (eg hello-testing-domain.com). You should see a result with your domain that is recommended.
5. search for an unavailable .blog FQDN (tim.blog), the first item should be your query but marked unavailable.

Fixes the first OC from: https://github.com/Automattic/wp-calypso/issues/41219
Fixes #41066

### CODE REVIEW

**Comment by yansern** [comment]
Okay I glanced through the code so my understanding is that if the FQDN is not in the list of suggestions, it is.. safe to say.. it is unavailable, yeah? And another PR to actually make the API request to find out why it is unavailable.

I smoke tested this and it works fine. :) I see domains that are bought are marked as unavailable.

File: packages/data-stores/package.json [hunk 1]
```diff
@@ -44,11 +44,13 @@
...
 		"tslib": "^2.1.0",
 		"utility-types": "^3.10.0",
+		"validator": "^13.5.2",
 		"wpcom-proxy-request": "^6.0.0"
 	},
 	"peerDependencies": {
 		"@wordpress/data": "^4",
-		"react": "^16.8"
+		"react": "^16.8",
+		"@types/validator": "^13.1.3"
```
Comment (StefanNieuwenhuis) [inline packages/data-stores/package.json]: In the root `package.json`, the types are added as `devDependencies`. Is there a specific reason to add them as `peerDependencies` here? And why?

File: packages/domain-picker/src/domain-picker/index.tsx [hunk 2]
```diff
@@ -383,9 +383,9 @@ const DomainPicker: FunctionComponent< Props > = ( {
 																suggestionRefs.current[ index ] = ref;
 															} }
 															key={ suggestion.domain_name }
-															isUnavailable={ ! isAvailable }
 															domain={ suggestion.domain_name }
 															cost={ suggestion.cost }
+															isUnavailable={ ! isAvailable || suggestion?.unavailable }
```
Comment (StefanNieuwenhuis) [inline packages/domain-picker/src/domain-picker/index.tsx]: I think it's better to include `suggestion?.unavailable` to `isAvailable` since it groups all logic to define an FQDN is available. WDYT?

File: packages/launch/src/utils.ts [hunk 3]
```diff
@@ -44,6 +44,9 @@ export const getDomainProduct = (
 	domain: DomainSuggestions.DomainSuggestion,
 	flow: string
 ): DomainProduct | undefined => {
+	if ( domain.unavailable ) {
```
Comment (StefanNieuwenhuis) [inline packages/launch/src/utils.ts:47]: Since they return the same, I think that we can merge both if-statements.

File: packages/data-stores/src/domain-suggestions/resolvers.ts [hunk 4]
```diff
@@ -68,20 +73,44 @@ export function* __internalGetDomainSuggestions( queryObject: DomainSuggestionQu
...
-	} ) );
+	// if the query is a FQDN and the results don't have it,
+	// this implies that the user is searching for an unavailable domain name
+	// TODO: query the availability endpoint to find the exact reason why it's unavailable
+	// all the possible responses can be found here https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/domains/registration/availability-messages.js#L40-L390
+	if (
+		validator.isFQDN( queryObject.query ) &&
+		suggestions &&
+		! suggestions.some( ( s ) => s.domain_name.toLowerCase() === queryObject.query )
+	) {
```
Comment (StefanNieuwenhuis) [inline packages/data-stores/src/domain-suggestions/resolvers.ts]: To make the code more comprehensible, we might want to consider refactoring this to a separate method like is done with `getAvailabilityURL`.

I think we don't have to check for suggestions to be truthy since we already did that on [line 76](https://github.com/Automattic/wp-calypso/pull/50998/files#diff-bb777e166faabc1fc42f24ab5ea72007101cb0f76eeea4a361a4aa5602fb29a5R76).

**Review by StefanNieuwenhuis — COMMENTED** [review]
Great job @alshakero! This is a huge leap in solving #41219 that's open since April 2020! 

I was able to run the tests successfully 🎉 

I noticed that you prepend the unavailable FQDN to the suggestions and I'm having a few concerns here since the data structure of an available & unavailable FQDN is different.

An unavailable FQDN consists only of a query + availability message while an available FQDN consists of (potentially) all props in the `DomainSuggestion` interface. 

This translates into the UI as well. We have to explain why an FQDN is unavailable, something we don't have to do for the available suggestions.

Maybe we want to split them up as in serving two separate objects. WDYT? 

Another thing I noticed is the following:

https://user-images.githubusercontent.com/12104589/111442123-51fdae80-8708-11eb-82e5-e5ba4c694532.mov

We still apply `cursor: hand` while hovering over the unavailable FQDN, rendering the illusion that it's still selectable.

File: packages/domain-picker/src/domain-picker/index.tsx (hunk 2 above)
Comment (alshakero) [reply packages/domain-picker/src/domain-picker/index.tsx]: This is theoretically true, but when you true to move it up the condition becomes quite dirty. 

```js
const isAvailable =
	( availabilityStatus
		? domainIsAvailableStatus?.indexOf( availabilityStatus ) > -1
		: true ) &&
	suggestion &&
	! suggestion.unavailable;
```

File: packages/launch/src/utils.ts (hunk 3 above)
Comment (alshakero) [reply packages/launch/src/utils.ts:47]: Let's be inspired by p7H4VZ-2Vx-p2 and keep them.

File: packages/data-stores/src/domain-suggestions/resolvers.ts (hunk 4 above)
Comment (alshakero) [reply packages/data-stores/src/domain-suggestions/resolvers.ts]: Great ideas.

File: packages/data-stores/package.json (hunk 1 above)
Comment (alshakero) [reply packages/data-stores/package.json]: Good point. I moved it to `devDependencies`.

**Comment by razvanpapadopol** [comment]
@alshakero does this PR fixes https://github.com/Automattic/wp-calypso/issues/41066 or at least a part of it?

File: packages/domain-picker/src/domain-picker/index.tsx (hunk 2 above)
Comment (StefanNieuwenhuis) [reply packages/domain-picker/src/domain-picker/index.tsx]: It might be a bit farfetched, but what if we declare 2 `const` and combine them?

```ts
const isDomainAvailable = availabilityStatus
		? domainIsAvailableStatus?.indexOf( availabilityStatus ) > -1
		: true;
const isSuggestionAvailable = suggestion && ! suggestion.unavailable;
const isAvailable = isDomainAvailable && isSuggestionAvailable;
```

**Comment by alshakero** [comment]
@razvanpapadopol yes it does fix https://github.com/Automattic/wp-calypso/issues/41066!

File: packages/data-stores/src/domain-suggestions/resolvers.ts [hunk 5]
```diff
@@ -68,20 +80,40 @@ export function* __internalGetDomainSuggestions( queryObject: DomainSuggestionQu
...
-		...( suggestion.raw_price &&
-			suggestion.currency_code && {
-				cost: getFormattedPrice( suggestion.raw_price, suggestion.currency_code ),
-			} ),
-	} ) );
+	// if the query is a FQDN and the results don't have it,
+	// this implies that the user is searching for an unavailable domain name
+	// TODO: query the availability endpoint to find the exact reason why it's unavailable
+	// all the possible responses can be found here https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/domains/registration/availability-messages.js#L40-L390
+	if ( suggestionsLackThisFQDN( suggestions, queryObject.query ) ) {
```
Comment (StefanNieuwenhuis) [inline packages/data-stores/src/domain-suggestions/resolvers.ts:94]: When the user searches for an unavailable domain, we mock a DomainSuggestion and prepend it to the list of suggestions.

File: packages/data-stores/src/domain-suggestions/resolvers.ts [hunk 6]
```diff
@@ -16,13 +17,23 @@ import {
...
+}
+
+function suggestionsLackThisFQDN( suggestions: DomainSuggestion[], domainName: string ) {
+	return (
+		validator.isFQDN( domainName ) &&
+		! suggestions.some( ( s ) => s.domain_name.toLowerCase() === domainName )
+	);
+}
+
+export const isAvailable = function* isAvailable( domainName: string ) {
```
Comment (StefanNieuwenhuis) [inline packages/data-stores/src/domain-suggestions/resolvers.ts:35]: This is called when the `isAvailable` selector in `packages/data-stores/src/domain-suggestions/selectors.ts` is called.

**Comment by StefanNieuwenhuis** [comment]
> I noticed that you prepend the unavailable FQDN to the suggestions and I'm having a few concerns here since the data structure of an available & unavailable FQDN is different.
> 
> An unavailable FQDN consists only of a query + availability message while an available FQDN consists of (potentially) all props in the `DomainSuggestion` interface.
> 
> This translates into the UI as well. We have to explain why an FQDN is unavailable, something we don't have to do for the available suggestions.
> 
> Maybe we want to split them up as in serving two separate objects. WDYT?

What do you think of the above? I think that we should address it in a separate issue since it requires quite some changes.

**Review by StefanNieuwenhuis — APPROVED** [review]
I've tested the changes and they work like a charm. Code changes LGTM! 

Can you please address https://github.com/Automattic/wp-calypso/pull/50998#issuecomment-810871489? Thanks 😊  

I'd say: 🚢 (maybe not via the Suez Canal...).

**Comment by alshakero** [comment]
Regarding [this](https://github.com/Automattic/wp-calypso/pull/50998#issuecomment-810871489) comment. 

That's how I started out my PR, you can see the [first commit](https://github.com/Automattic/wp-calypso/pull/50998/commits/a653e019438d17117c846dfc0daa98fd4557c906), but it grew in complexity so bad, I realized adding the unavailable flag is much easier and more backward compatible (because `unavailable` only needs to exist when it's needed. But `available` needs to be added to every available suggestion).

File: packages/data-stores/src/domain-suggestions/resolvers.ts (hunk 6 above)
Comment (alshakero) [reply packages/data-stores/src/domain-suggestions/resolvers.ts:35]: Yes

**Comment by StefanNieuwenhuis** [comment]
> Regarding [this](https://github.com/Automattic/wp-calypso/pull/50998#issuecomment-810871489) comment.
> 
> That's how I started out my PR, you can see the [first commit](https://github.com/Automattic/wp-calypso/pull/50998/commits/a653e019438d17117c846dfc0daa98fd4557c906), but it grew in complexity so bad, I realized adding the unavailable flag is much easier and more backward compatible (because `unavailable` only needs to exist when it's needed. But `available` needs to be added to every available suggestion).

Thanks for elaborating! I already thought this would become complicated very easily and this solution is indeed much easier and more backward compatible.

File: packages/domain-picker/src/components/index.tsx [hunk 7]
```diff
@@ -160,9 +160,14 @@ const DomainPicker: FunctionComponent< Props > = ( {
 		retryRequest: retryDomainSuggestionRequest,
 	} = useDomainSuggestions( domainSearch.trim(), quantityExpanded, domainCategory, locale ) || {};
 
-	const domainSuggestions = allDomainSuggestions?.slice(
-		existingSubdomain ? 1 : 0,
-		isExpanded ? quantityExpanded : quantity
+	// don't list the already existing free domain as part of the suggestions
+	const domainSuggestions = allDomainSuggestions
+		?.filter( ( suggestion ) => suggestion.domain_name !== existingSubdomain?.domain_name )
+		.slice( 0, isExpanded ? quantityExpanded : quantity );
```
Comment (razvanpapadopol) [inline packages/domain-picker/src/components/index.tsx:166]: @StefanNieuwenhuis @alshakero 
Was this change needed in the context of this PR or it was just cleanup/refactoring?
I'm asking because it has introduced a regression in Focused Launch. See p1617903103117300-slack-C013QHLF28Y

Here is a fix (reverting this change): https://github.com/Automattic/wp-calypso/pull/51806

Comment (StefanNieuwenhuis) [reply packages/domain-picker/src/components/index.tsx:166]: I did a little debugging for `domainSuggestions`:
1. filter out all suggestions with a `domain_name` equal to the existing subdomain.
2. slice the filtered list to the predefined length.

While testing (stefannew.wordpress.com), my domain suggestions are:
* free subdomain: `stefannew319442470`.
* `stefan-new.com`.
* `stefannew.life`.
* `stefannew.blog`.

The existing subdomain is: `stefannew` so the free subdomain suggestion isn't filtered from the list resulting in the regression!

The previous implementation didn't care about `domain_name` and **always** cuts out the free subdomain when available. 

### Conclusion
* **Current implementation**: More specific but causes regressions when the names of the free subdomains don't match.
* **Previous implementation**:  Always cuts out the first suggestion when a free subdomain is available. 

It seems to me that both approaches try to achieve the same so I think it's safe to assume that this change was made to improve the domain suggestions filtering logic.

Since the latter doesn't cause regressions (for now) and provides a solution for p1617903103117300-slack-C013QHLF28Y, I think it's okay to apply the revert (#51806).

Comment (alshakero) [reply packages/domain-picker/src/components/index.tsx:166]: Yes it was needed in this PRs context. Because this PR broke two assumptions we used to make:

1. The free domain is always the first suggestion, so we just shifted the array to remove it. This assumption became false when we added the unavailable suggestion at the top when users search for an FQDN.
2. That the second item of the array is the recommended domain (the first paid domain).

This change was an attempt at:

1. Find the free domain no matter where it is.
2. Find the first occurring paid domain in the suggestion and suggest it as "recommended".

Comment (StefanNieuwenhuis) [reply packages/domain-picker/src/components/index.tsx:166]: Thanks for elaborating!
